designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,308 @@
+# Local Debugging Experience for SAM Applications
+
+Current Status: Not Implemented
+
+## Introduction
+
+The AWS Toolkit enhances the SAM Application development experience by integrating local debug functionality into VS Code. This document outlines the available functionality.
+
+While this document's main focus is on debugging capabilities in the toolkit, there are places where the experience around invoking without the debugger (aka "running") is also discussed.
+
+Each programming language (and corresponding Lambda Runtimes) requires Toolkit support for debugging features to work. A limited selection of programming languages are supported in the Toolkit.
+
+### Terminology
+
+#### CodeLens
+
+CodeLenses are visual decorators anchored to a document location. They are used to convey information and/or provide links that trigger an action. They are a presentation-only mechanic and do not reside within a file. Additional information and examples about CodeLenses can be found [on the VS Code blog](https://code.visualstudio.com/blogs/2017/02/12/code-lens-roundup).
+
+#### Debug Configuration
+
+Debug Configurations are JSON entries within the `.vscode/launch.json` file optionally located in the root of each VS Code workspace. These are user managed, and define what programs can be debugged. Presing F5 (or the Debug button) starts a Debugging session for the currently selected Debug Configuration. VS Code extensions increase VS Code's debugging capablities by implementing Debug Configuration types.
+
+More information about VS Code Debugging can be found [in the VS Code Documentation](https://code.visualstudio.com/docs/editor/debugging).
+
+#### SAM Template
+
+A SAM Template defines a Serverless Application's resources, and supporting code. This is used by the SAM CLI to build, run, package, and deploy the Application.
+
+Additional information about SAM can be found at:
+
+-   [SAM Homepage](https://aws.amazon.com/serverless/sam/)
+-   [What Is the AWS Serverless Application Model (AWS SAM)?](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/what-is-sam.html)
+-   [SAM CLI GitHub Repo](https://github.com/awslabs/aws-sam-cli)
+
+## Overview
+
+The toolkit supports the following scenarios for Locally Running and Debugging code using the Serverless Application Model:
+
+-   Invoking SAM Template Lambda Function Resources
```

Comment:
nit: is a "SAM Template Lambda Function Resource" really a proper-noun? (i.e. does every letter need to be title-cased?)

Reply:
I wasn't consistent in the doc on this. I tended to capitalize Resource because of the SAM Template spec, but looking at https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/what-is-sam.html and https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-specification-template-anatomy.html I think resource shouldn't be considered capitalized. I'll do a pass over this.